### PR TITLE
Lots of changes to the LAS layer, including support for variable-sized chunks

### PR DESCRIPTION
**Pull request by abellgithub**

### CODE REVIEW

File: cpp/lazperf/lazperf.cpp [hunk 1]
```diff
@@ -525,6 +525,9 @@ las_compressor::ptr build_las_compressor(OutputCb cb, int format, size_t ebCount
         break;
     case 8:
         compressor.reset(new point_compressor_8(cb, ebCount));
+        break;
+    default:
+        std::cerr << "Invalid/unsupported format " << format << ".\n";
```
Comment (hobu) [inline cpp/lazperf/lazperf.cpp:530]: Should this throw or do some other kind of error handling?

File: cpp/lazperf/vlr.hpp [hunk 2]
```diff
@@ -113,12 +127,50 @@ struct eb_vlr : public vlr
...
+    virtual ~wkt_vlr();
+
+    static wkt_vlr create(std::istream& in, int byteSize);
+    void read(std::istream& in, int byteSize);
+    void write(std::ostream& out) const;
+    virtual size_t size() const;
+    virtual vlr_header header() const;
+};
+
+struct LAZPERF_EXPORT copc_vlr : public vlr
```
Comment (hobu) [inline cpp/lazperf/vlr.hpp:158]: Does this belong here? I don't have a strong opinion, but I thought we stated that COPC library convenience wasn't going to be implemented in laz-perf.

**Review by hobu — APPROVED** [review]

File: cpp/lazperf/vlr.hpp (hunk 2 above)
Comment (abellgithub) [reply cpp/lazperf/vlr.hpp:158]: It simplifies my life, so...

File: cpp/lazperf/lazperf.cpp (hunk 1 above)
Comment (abellgithub) [reply cpp/lazperf/lazperf.cpp:530]: Not sure, which is why it just prints an error. The caller gets a nullptr so should be able to detect and handle. 🤷

Comment (hobu) [reply cpp/lazperf/lazperf.cpp:530]: ok, but libraries shouldn't just print random stuff to stderr/stdout IMO.

File: cpp/lazperf/header.cpp [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+bool base_header::compressed() const
+{
+    return point_format_id & 0x80;
+}
+
+size_t base_header::sizeFromVersion() const
+{
+    size_t size = 0;
+    if (version.minor == 2)
+        size = 227;      // 1.2 size
```
[the remark this answers is not in the document]
Comment (abellgithub) [reply cpp/lazperf/header.cpp:88]: Yes.

File: cpp/lazperf/header.cpp [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+    else if (version.minor == 3)
+        size = 235;     // 1.3 size
+    else if (version.minor == 4)
+        size = 375;     // 1.4 size
+    else
+        std::cerr << "Invalid minor version " << (int)version.minor <<
```
[the remark this answers is not in the document]
Comment (abellgithub) [reply cpp/lazperf/header.cpp:94]: Agree.

File: cpp/lazperf/writers.hpp [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+protected:
+    std::unique_ptr<Private> p_; 
+};
+
+class named_file : public basic_file
+{
+    struct Private;
+
+public:
+    struct config
```
Comment (CCInc) [inline cpp/lazperf/writers.hpp:60]: This should be `LAZPERF_EXPORT`

File: cpp/lazperf/vlr.hpp [hunk 6]
```diff
@@ -113,12 +127,50 @@ struct eb_vlr : public vlr
...
+{
+public:
+    uint64_t root_hier_offset;
+    uint64_t root_hier_size;
+    uint64_t span;
```
Comment (CCInc) [inline cpp/lazperf/vlr.hpp:163]: in the spec span is a `int64_t`

**Comment by CCInc** [comment]
Can we expose the `header14` object through the `readers` api? so I can access the EVLR offset/size.

**Comment by abellgithub** [comment]
It's available as lazperf::header12 with the function header().

**Comment by CCInc** [comment]
I can't access the EVLR attributes with `header12`.

**Comment by abellgithub** [comment]
This is historic, but I changed it. Note that while the data will be accessible, you must check the version to make sure that it's valid.

**Pull request merged**